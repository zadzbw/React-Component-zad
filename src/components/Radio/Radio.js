/**
 * Created by zad on 17/4/17.
 */
import React, {PureComponent, PropTypes} from 'react';
import classNames from 'classnames';

export const radioPrefix = 'zad-radio';

export default class Radio extends PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      checked: 'checked' in this.props ? this.props.checked : this.getInitChecked(),
    };
  }

  static propTypes = {
    disabled: PropTypes.bool,
    checked: PropTypes.bool,
    defaultChecked: PropTypes.bool,
    onChange: PropTypes.func,
    wrapperClassName: PropTypes.string,
  };

  static defaultProps = {
    disabled: false,
    onChange: () => undefined,
  };

  componentWillReceiveProps(nextProps) {
    if ('checked' in nextProps) {
      this.setState({
        checked: nextProps.checked,
      });
    }
  }

  getInitChecked = () => {
    const props = this.props;
    if ('defaultChecked' in props) {
      return props.defaultChecked;
    }
    return false;
  };

  changeChecked = (e) => {
    const {onChange, disabled} = this.props;
    if (disabled) {
      return false;
    }
    if (!('checked' in this.props)) {
      this.setState({
        checked: e.target.checked,
      });
    }
    onChange(e);
  };

  render() {
    const {disabled, wrapperClassName, children} = this.props;
    const {checked} = this.state;

    const wrapperClass = classNames(wrapperClassName, {
      [`${radioPrefix}-wrapper`]: true,
      [`${radioPrefix}-wrapper-disabled`]: disabled,
    });
    const radioClass = classNames(radioPrefix, {
      [`${radioPrefix}-checked`]: checked,
    });

    return (
      <label className={wrapperClass}>
        <span className={radioClass}>
          <input type="radio"
                 className={`${radioPrefix}-input`}
                 checked={checked}
                 onChange={this.changeChecked}
          />
          <span className={`${radioPrefix}-mock`}/>
        </span>
        <span className={`${radioPrefix}-name`}>
          {children}
        </span>
      </label>
    );
  }
}
