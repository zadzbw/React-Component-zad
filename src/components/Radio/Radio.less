@import "../../style/variables";
@import "../../style/mixins/index";

@radio-class: zad-radio;
@radio-group-class: ~"@{radio-class}-group";

.@{radio-class} {
  position: relative;
  display: inline-block;
  line-height: 1;
  vertical-align: text-bottom; // 保证假的radio与name在同一水平线上

  &-wrapper {
    cursor: pointer;
    font-size: @font-size-sm;
    display: inline-block;

    &:hover {
      .@{radio-class}-mock {
        border-color: @primary-color;
      }
    }

    &-disabled {
      cursor: not-allowed;
      .@{radio-class}-name {
        color: @disabled-color;
      }

      .@{radio-class}-mock {
        border-color: @border-color-base !important; // 边框色
        background-color: lighten(@border-color-base, 10%) !important; // 背景色

        &:after {
          animation-name: none;
          background-color: @border-color-base; // 圆点色
        }
      }
    }
  }

  &-input:focus + &-mock {
    border-color: @primary-color;
  }

  &-input {
    position: absolute;
    left: 0;
    top: 0;
    bottom: 0;
    right: 0;
    width: 100%;
    height: 100%;
    z-index: 1;
    opacity: 0;
  }

  // 假的radio
  &-mock {
    position: relative;
    top: 0;
    left: 0;
    display: block;
    width: 14px;
    height: 14px;
    border: 1px solid @border-color-base;
    border-radius: 50%;
    background-color: #fff;
    transition: all .3s;
    z-index: 2;

    // 创建一个block，利用background画出圆点
    &:after {
      transform: scale(0);
      position: absolute;
      display: table;
      left: 3px;
      top: 3px;
      width: 6px;
      height: 6px;
      border-radius: 50%;
      background-color: @primary-color;
      content: ' ';
      transition: all .1s @ease-in-back;
    }
  }

  &-name {
    padding-left: 8px;
    padding-right: 8px;
  }

  &-checked {
    .@{radio-class}-mock:after {
      transform: scale(1);
      transition: all .2s @ease-out-back .1s;
    }
  }

  &-checked {
    .@{radio-class}-mock {
      border-color: @primary-color;
    }
  }
}

.@{radio-group-class} {
  font-size: @font-size-sm;
  &-item {
    display: inline-block;
    &:not(:last-child) {
      margin-right: 8px;
    }
  }
}
