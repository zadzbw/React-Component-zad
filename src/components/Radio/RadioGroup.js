/**
 * Created by zad on 17/4/17.
 */
import React, {PureComponent, PropTypes, createElement} from 'react';
import classNames from 'classnames';
import _omit from 'lodash/omit';
import _isString from 'lodash/isString';
import _isFunction from 'lodash/isFunction';
import Radio, {radioPrefix} from './Radio';

const radioGroupPrefix = `${radioPrefix}-group`;

export default class RadioGroup extends PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      value: 'value' in this.props ? this.props.value : this.props.defaultValue,
    };
  }

  static propTypes = {
    value: PropTypes.string,
    defaultValue: PropTypes.string,
    options: PropTypes.array,
    onChange: PropTypes.func,
  };

  static defaultProps = {
    defaultValue: '',
    options: [],
    onChange: () => undefined,
  };

  componentWillReceiveProps(nextProps) {
    if ('value' in nextProps) {
      this.setState({
        value: nextProps.value || '',
      });
    }
  }

  changeChecked = (v) => {
    return (event) => {
      // 这里的event是由Radio的onChange函数传递的
      const {value} = this.state;
      const isChange = value !== v;
      if (isChange) {
        if (!('value' in this.props)) {
          this.setState({
            value: v,
          });
        }
        const {onChange} = this.props;
        if (_isFunction(onChange)) {
          onChange({
            event,
            value: v,
          });
        }
      }
    };
  };

  getOptions = () => {
    const {options} = this.props;
    return options.map((item) => {
      if (_isString(item)) {
        return {name: item, value: item, disabled: false};
      }
      return item;
    });
  };

  getRadios = () => {
    const options = this.getOptions();
    const {value} = this.state;
    return options.map((option, i) => {
      return createElement(Radio, {
        key: `group-radio-${i}`,
        disabled: option.disabled,
        children: option.name,
        checked: value === option.value,
        wrapperClassName: `${radioGroupPrefix}-item`,
        onChange: this.changeChecked(option.value),
      });
    });
  };

  render() {
    const {className} = this.props;
    const restProps = _omit(this.props, [
      'className',
      'value',
      'defaultValue',
      'options',
      'onChange',
    ]);
    const groupClass = classNames(radioGroupPrefix, className);
    return (
      <div className={groupClass} {...restProps}>
        {this.getRadios()}
      </div>
    );
  }
}
