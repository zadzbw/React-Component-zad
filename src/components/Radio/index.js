/**
 * Created by zad on 17/4/17.
 */
import './Radio.less';

import Radio from './Radio';


export default Radio;
