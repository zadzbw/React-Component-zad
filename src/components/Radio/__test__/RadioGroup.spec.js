/**
 * Created by zad on 17/4/17.
 */
import React from 'react';
import {shallow} from 'enzyme';
import {shallowToJson} from 'enzyme-to-json';
import Radio from '../Radio';
import RadioGroup from '../RadioGroup';

describe('Radio Test', () => {
  it('controlled component', () => {
    class Test extends React.Component {
      constructor(props) {
        super(props);
        this.state = {radio_group: 'other'};
      }

      testCheckGroup = ({value}) => {
        this.setState({
          radio_group: value,
        })
      };

      render() {
        return (
          <RadioGroup
            options={[
              {name: '男性', value: 'male'},
              {name: '女性', value: 'female'},
              {name: '其他', value: 'other'},
            ]}
            value={this.state.radio_group}
            onChange={this.testCheckGroup}
          />
        );
      }
    }

    const wrapper = shallow(<Test/>);

    expect(wrapper.state('radio_group')).toEqual('other');
    expect(wrapper.shallow().find(Radio).at(2).prop('checked')).toBe(true);

    expect(wrapper.shallow().find(Radio).at(0).simulate('change')); // male

    expect(wrapper.state('radio_group')).toEqual('male');
    expect(wrapper.shallow().find(Radio).at(0).prop('checked')).toBe(true);

    expect(shallowToJson(wrapper.shallow())).toMatchSnapshot();
  });

  it('uncontrolled component', () => {
    const wrapper = shallow(
      <RadioGroup
        options={[
          {name: '男性', value: 'male'},
          {name: '女性', value: 'female'},
          {name: '其他', value: 'other'},
        ]}
        defaultValue={'female'}
      />
    );

    expect(wrapper.state('value')).toEqual('female');
    expect(wrapper.find(Radio).at(1).prop('checked')).toBe(true);
    expect(wrapper.shallow().find(Radio).at(2).simulate('change'));
    expect(wrapper.state('value')).toEqual('other');
    expect(wrapper.find(Radio).at(2).prop('checked')).toBe(true);

    expect(shallowToJson(wrapper.shallow())).toMatchSnapshot();
  });

  it('with disabled', () => {
    const wrapper = shallow(
      <RadioGroup
        options={[
          {name: '男性', value: 'male'},
          {name: '女性', value: 'female', disabled: true},
          {name: '其他', value: 'other'},
        ]}
        defaultValue={'female'}
      />
    );

    expect(wrapper.find(Radio).at(1).prop('checked')).toBe(true);

    expect(wrapper.find(Radio).at(0).prop('disabled')).toBe(false);
    expect(wrapper.find(Radio).at(1).prop('disabled')).toBe(true);
    expect(wrapper.find(Radio).at(2).prop('disabled')).toBe(false);

    expect(shallowToJson(wrapper.shallow())).toMatchSnapshot();
  });
});
