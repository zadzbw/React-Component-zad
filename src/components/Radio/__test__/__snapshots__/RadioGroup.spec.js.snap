exports[`Radio Test controlled component 1`] = `
<div
  className="zad-radio-group">
  <Radio
    checked={true}
    disabled={false}
    onChange={[Function]}
    wrapperClassName="zad-radio-group-item">
    男性
  </Radio>
  <Radio
    checked={false}
    disabled={false}
    onChange={[Function]}
    wrapperClassName="zad-radio-group-item">
    女性
  </Radio>
  <Radio
    checked={false}
    disabled={false}
    onChange={[Function]}
    wrapperClassName="zad-radio-group-item">
    其他
  </Radio>
</div>
`;

exports[`Radio Test uncontrolled component 1`] = `
<div
  className="zad-radio-group">
  <Radio
    checked={false}
    disabled={false}
    onChange={[Function]}
    wrapperClassName="zad-radio-group-item">
    男性
  </Radio>
  <Radio
    checked={false}
    disabled={false}
    onChange={[Function]}
    wrapperClassName="zad-radio-group-item">
    女性
  </Radio>
  <Radio
    checked={true}
    disabled={false}
    onChange={[Function]}
    wrapperClassName="zad-radio-group-item">
    其他
  </Radio>
</div>
`;

exports[`Radio Test with disabled 1`] = `
<div
  className="zad-radio-group">
  <Radio
    checked={false}
    disabled={false}
    onChange={[Function]}
    wrapperClassName="zad-radio-group-item">
    男性
  </Radio>
  <Radio
    checked={true}
    disabled={true}
    onChange={[Function]}
    wrapperClassName="zad-radio-group-item">
    女性
  </Radio>
  <Radio
    checked={false}
    disabled={false}
    onChange={[Function]}
    wrapperClassName="zad-radio-group-item">
    其他
  </Radio>
</div>
`;
